static/js/narrow.js

```diff
@@ -287,79 +291,84 @@ exports.maybe_add_local_messages = function (opts) {
+                id_info.search_id,
+                unread_info.msg_id
+            );
         } else if (unread_info.flavor === 'not_found') {
-            // If we didn't find any unread messages, then we
-            // generally want to go the last message in the list,
-            // but only if we've fetched the newest messages
-            // and the narrow's not empty locally.
-            if (message_list.all.fetch_status.has_found_newest()) {
-                // Load messages now and upgrade our strategy
-                // if we find at least one message.
-                if (load_local_messages(msg_data)) {
-                    var last_msg = msg_data.last();
-                    select_strategy = {
-                        flavor: 'exact',
-                        msg_id: last_msg.id,
-                    };
-
-                    return select_strategy;
+            id_info.final_select_id = id_info.search_id;
+        } else {
+            // if we can't compute a next unread id, just
+            // return without setting local_select_id to be
+            // on the safe side
+            id_info.final_select_id = id_info.search_id;
+            return;
+        }
+
+        if (narrow_state.get_current_filter().can_apply_locally()) {
+            if (load_local_messages(msg_data)) {
+                if (msg_data.get(id_info.final_select_id)) {
+                    id_info.local_select_id = id_info.final_select_id;
                 }
             }
         }
+        return;
     }
 
-    // Do not make this an else-if...the block above could have
-    // changed select_strategy.
-    if (select_strategy.flavor === 'exact') {
-        var msg_to_select = message_list.all.get(select_strategy.msg_id);
-        if (msg_to_select !== undefined) {
+    unread_info = narrow_state.get_first_unread_info();
+
+    if (unread_info.flavor === 'found') {
+        id_info.final_select_id = min_defined(
+            id_info.previous_id,
+            unread_info.msg_id
+        );
+        if (load_local_messages(msg_data)) {
+            if (msg_data.get(id_info.final_select_id)) {
+                id_info.local_select_id = id_info.final_select_id;
+            }
+        }
+        return;
+    } else if (unread_info.flavor === 'not_found') {
+        // If we didn't find any unread messages, then we
+        // generally want to go the last message in the list,
+        // but only if we've fetched the newest messages
+        // and the narrow's not empty locally.
+        if (message_list.all.fetch_status.has_found_newest()) {
+            // Load messages now and upgrade our strategy
+            // if we find at least one message.
             if (load_local_messages(msg_data)) {
-                return select_strategy;
+                var last_msg = msg_data.last();
+                id_info.final_select_id = min_defined(
+                    last_msg.id,
+                    id_info.previous_id
+                );
```

Comment:
I think this block isn't quite right: we might be looking at a super-old message in a topic narrow, and trying to move to the corresponding stream narrow.  In that event, I think this will end up narrowing us to the oldest message in `all_msg_list` matching the narrow, just because it'll locally render a message list that only has the `all_msg_list` range of messages.  And then we'll end up landing on that oldest message through `select_id`, rather than going to the server?  

Maybe I'm missing something.